zipkin-server/src/main/java/zipkin2/server/internal/health/ZipkinHealthController.java

```diff
@@ -0,0 +1,153 @@
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static zipkin2.server.internal.ZipkinServerConfiguration.MEDIA_TYPE_ACTUATOR;
+import static zipkin2.server.internal.health.ComponentHealth.STATUS_DOWN;
+import static zipkin2.server.internal.health.ComponentHealth.STATUS_UP;
+
+public class ZipkinHealthController {
+  static final JsonFactory JSON_FACTORY = new JsonFactory();
+
+  final List<Component> components;
+
+  ZipkinHealthController(List<Component> components) {
+    this.components = components;
+  }
+
+  @Get("/actuator/health")
+  public CompletableFuture<HttpResponse> getActuatorHealth(ServiceRequestContext ctx) {
+    return health(ctx, MEDIA_TYPE_ACTUATOR);
+  }
+
+  @Get("/health")
+  public CompletableFuture<HttpResponse> getHealth(ServiceRequestContext ctx) {
+    return health(ctx, MediaType.JSON_UTF_8);
+  }
+
+  CompletableFuture<HttpResponse> health(ServiceRequestContext ctx, MediaType mediaType) {
+    CompletableFuture<HttpResponse> responseFuture = new CompletableFuture<>();
+    ctx.setRequestTimeoutHandler(() -> {
+      try {
+        String healthJson = writeJsonError("Timed out computing health status. "
+          + "This often means your storage backend is unreachable.");
+        responseFuture.complete(newHealthResponse(STATUS_DOWN, mediaType, healthJson));
+      } catch (Throwable e) {
+        // Shouldn't happen since we serialize to an array.
+        responseFuture.completeExceptionally(e);
+      }
+    });
+
+    ctx.blockingTaskExecutor().execute(() -> {
+      try {
+        responseFuture.complete(newHealthResponse(componentHealths(components), mediaType));
+      } catch (Throwable e) {
+        responseFuture.completeExceptionally(e);
+      }
+    });
+    return responseFuture;
+  }
+
+  static List<ComponentHealth> componentHealths(List<Component> components) {
+    return components.stream()
+      .parallel().map(ComponentHealth::ofComponent)
+      .collect(Collectors.toList());
+  }
+
+  static HttpResponse newHealthResponse(List<ComponentHealth> healths, MediaType mediaType)
+    throws IOException {
+
+    String overallStatus = STATUS_UP;
+    for (ComponentHealth health : healths) {
+      if (health.status.equals(STATUS_DOWN)) overallStatus = STATUS_DOWN;
+    }
+
+    String healthJson = writeJson(overallStatus, healths);
+    return newHealthResponse(overallStatus, mediaType, healthJson);
+  }
+
+  static HttpResponse newHealthResponse(String overallStatus, MediaType mediaType,
+    String healthJson) {
+    byte[] body = healthJson.getBytes(UTF_8);
+    int code = overallStatus.equals(STATUS_UP) ? 200 : 503;
+    ResponseHeaders headers = ResponseHeaders.builder(code)
+      .contentType(mediaType)
+      .setInt(CONTENT_LENGTH, body.length).build();
+    return HttpResponse.of(AggregatedHttpResponse.of(headers, HttpData.wrap(body)));
+  }
+
+  static String writeJsonError(String error) throws IOException {
+    StringWriter writer = new StringWriter();
```

Comment:
Could consider `ByteArrayOutputStream` since it immediately becomes bytes anyways